core/src/test/scala/unit/kafka/server/CreateTopicsRequestTest.scala

```diff
@@ -0,0 +1,212 @@
+      else
+        details.numPartitions
+
+      val replication = if (!details.replicasAssignments.isEmpty)
+        details.replicasAssignments.asScala.head._2.size
+      else
+        details.replicationFactor
+
+      assertNotNull("The topic should be created", metadataForTopic)
+      assertEquals("The topic should have the correct number of partitions", partitions, metadataForTopic.partitionMetadata.size)
+      assertEquals("The topic should have the correct replication factor", replication, metadataForTopic.partitionMetadata.asScala.head.replicas.size)
+    }
+  }
+
+  @Test
+  def testErrorCreateTopicsRequests() {
+    val timeout = 10000
+    val existingTopic = "existing-topic"
+    TestUtils.createTopic(zkUtils, existingTopic, 1, 1, servers)
+
+    // Basic
+    validateErrorCreateTopicsRequests(new CreateTopicsRequest(Map(existingTopic -> new CreateTopicsRequest.TopicDetails(1, 1.toShort)).asJava, timeout),
+      Map(existingTopic -> Errors.TOPIC_ALREADY_EXISTS))
+    validateErrorCreateTopicsRequests(new CreateTopicsRequest(Map("error-partitions" -> new CreateTopicsRequest.TopicDetails(-1, 1.toShort)).asJava, timeout),
+      Map("error-partitions" -> Errors.INVALID_PARTITIONS))
+    validateErrorCreateTopicsRequests(new CreateTopicsRequest(Map("error-replication" -> new CreateTopicsRequest.TopicDetails(1, (numBrokers + 1).toShort)).asJava, timeout),
+      Map("error-replication" -> Errors.INVALID_REPLICATION_FACTOR))
+    val invalidConfig = Map("not.a.property" -> "error").asJava
+    validateErrorCreateTopicsRequests(new CreateTopicsRequest(Map("error-config" -> new CreateTopicsRequest.TopicDetails(1, 1.toShort, invalidConfig)).asJava, timeout),
+      Map("error-config" -> Errors.INVALID_CONFIG))
+    val invalidAssignments = replicaAssignmentToJava(Map(0 -> List(0, 1), 1 -> List(0)))
+    validateErrorCreateTopicsRequests(new CreateTopicsRequest(Map("error-assignment" -> new CreateTopicsRequest.TopicDetails(invalidAssignments)).asJava, timeout),
+      Map("error-assignment" -> Errors.INVALID_REPLICA_ASSIGNMENT))
+
+    // Partial
+    validateErrorCreateTopicsRequests(
+      new CreateTopicsRequest(Map(
+        existingTopic -> new CreateTopicsRequest.TopicDetails(1, 1.toShort),
+        "partial-partitions" -> new CreateTopicsRequest.TopicDetails(-1, 1.toShort),
+        "partial-replication" -> new CreateTopicsRequest.TopicDetails(1, (numBrokers + 1).toShort),
+        "partial-none" -> new CreateTopicsRequest.TopicDetails(1, 1.toShort)).asJava, timeout),
+      Map(
+        existingTopic -> Errors.TOPIC_ALREADY_EXISTS,
+        "partial-partitions" -> Errors.INVALID_PARTITIONS,
+        "partial-replication" -> Errors.INVALID_REPLICATION_FACTOR,
+        "partial-none" -> Errors.NONE
+      )
+    )
+    validateTopicExists("partial-none")
+
+    // Timeout
+    validateErrorCreateTopicsRequests(new CreateTopicsRequest(Map("error-timeout" -> new CreateTopicsRequest.TopicDetails(10, 3.toShort)).asJava, 1),
```

Comment:
The section of code I am testing is AdminManager. The gist of the code is:

```scala
// 2. if timeout <= 0 or no topics can proceed return immediately
if (timeout <= 0 || !metadata.exists(_.error == Errors.NONE)) {
   val results = ...
   responseCallback(results)
} else {
   // 3. else pass the assignments and errors to the delayed operation and set the keys
   val delayedCreate = ...
   val delayedCreateKeys = createInfo.keys.map(new TopicKey(_)).toSeq
   // try to complete the request immediately, otherwise put it into the purgatory
   topicPurgatory.tryCompleteElseWatch(delayedCreate, delayedCreateKeys)
}
```

Its the difference between just setting the timeout code and returning immediately vs putting it into purgatory and letting purgatory cause the timeout. I want to be sure we are testing both.